src/sim.py

```diff
@@ -71,46 +72,68 @@ def DepotsHaveNodes(router, depots, search_radius_m=1000):
   return good
 
 
+def generateParams(**kwargs):
+  """
+  Generate a Dispatch parameters object
+
+  """
+  #TODO enforce types
+  # simulator parameters
+  params = dispatch.Parameters()
+  params.battery_cap_kwh       = kwargs.get('battery_cap_kwh',      200)        #kWh
+  params.nondepot_charger_rate = kwargs.get('nondepot_charger_rate',  500)      #kW
+# params.nondepot_charger_den  = kwargs.get(street_charger_den, 0.25)           #ratio TODO implement support   
+  params.depot_charger_rate    = kwargs.get('depot_charger_rate',   125)        #kW
+  params.bus_cost              = kwargs.get('bus_cost',             500_000)    #dollars
+  params.battery_cost_per_kwh  = kwargs.get('battery_cost_per_kwh', 100)        #dollars
+  params.depot_charger_cost    = kwargs.get('depot_charger_cost',   50_000)     #dollars
+  params.nondepot_charger_cost = kwargs.get('nondepot_charger_cost',  600_000)  #dollars
+  params.kwh_per_km            = kwargs.get('kwh_per_km',           1.2)        #kWh_per_km
+  params.chargers_per_depot    = kwargs.get('chargers_per_depot',   1)          #TODO: Bad default
+  # optimizer parameters
+  params.generations           = kwargs.get('generations',          [ 50,  50]) #,1000]
+  params.mutation_rate         = kwargs.get('mutation_rate',        [0.1,0.05]) #,0.01]
+  params.keep_top              = kwargs.get('keep_top',             [  5,   5]) #,   5]
+  params.spawn_size            = kwargs.get('spawn_size',           [100, 100]) #,  50]
+  params.restarts              = kwargs.get('restarts',             1)
+  params.seed                  = kwargs.get('seed',                 0) #Initialize differently each time
+  print('Parameter Kwargs',kwargs,'\n\n')
+  return params
+
+def simulate(input_prefix,
+             osm_data,
+             depots_filename,
+             output_filename,
+             parameters=None
+            ):
+  """ TODO Performs an optimized simulation
+      parameters, dict of simulation and optimizer parameters to be used, 
+  """
 
-def simulate(
-  input_prefix,
-  osm_data,
-  depots_filename,
-  output_filename
-):
   print("Parsing OSM data into router...")
   router = dispatch.Router(osm_data)
 
+
   trips      = pd.read_csv(f"{input_prefix}_trips.csv")
   stops      = pd.read_csv(f"{input_prefix}_stops.csv")
   stop_times = pd.read_csv(f"{input_prefix}_stop_times.csv")
   depots     = pd.read_csv(depots_filename)
 
+  #TODO error handling for malformed dict, maybe should be in generateParams()?
+  if parameters is not None:
+    params = generateParams(**parameters)
```

Comment:
Since `generateParams` uses `get`, there's no need for an `if` statement here.

Reply:
Will a function gracefully execute if its passed an input that is not present?

Reply:
If it has a default argument, and you seem to have written this function so that passing `parameter=None` will results in an appropriate call to the Parameters object generator.